torch/csrc/jit/codegen/cuda/arith.cpp

```diff
@@ -1358,17 +1415,29 @@ TensorView* gather(
     const auto window_dim = window_shape[i];
     const auto pad_left = pad_width[i][0];
     const auto pad_right = pad_width[i][1];
+    // This may be over-conservative
     TORCH_INTERNAL_ASSERT(inp_axis->start()->isZeroInt());
+    const auto inp_stop_offset = inp_axis->stopOffset()->getInt();
+    TORCH_INTERNAL_ASSERT(
+        inp_stop_offset.has_value(),
+        "Dynamic stop offset not supported: ",
+        inp_axis);
+    const auto extent_adjustment = window_dim - 1 - pad_left - pad_right;
+    TORCH_CHECK(
+        extent_adjustment >= 0,
+        "Invalid gather window and padding as output extent would be larger than input.",
+        " Window: ",
+        window_dim,
+        ". Padding left: ",
+        pad_left,
+        ". Padding right: ",
+        pad_right);
+    const auto out_stop_offset = inp_stop_offset.value() + extent_adjustment;
     Val* out_axis_dim = nullptr;
-    const auto extent_adjustment = -(-window_dim + 1 + pad_left + pad_right);
-    out_axis_dim = extent_adjustment == 0
-        ? inp_axis->extent()
-        : sub(inp_axis->extent(), new Int(extent_adjustment));
-    // TODO: out_axis_dim is assumed to be the same as the extent of
-    // the input domain. Throw an error if it isn't the case.
     out_root_domains.push_back(new IterDomain(
         new Int(0),
```

Comment:
Why does both padding sides go into extent and left padding doesn't go into the start offset?

Reply:
This depends on which cell in a window is considered the base. If the left-most cell is considered the base, even if there's no left padding, the resulting output tensor should have zero start-offset. This is what I decided to do.

Alternatively, it's possible to choose a center cell as the base. For example, for an window of extent 3, we could pick the center one as the base, then there would be no output value defined for at offset 0, so the start offset would be 1.

Either should be fine for correctness, but there could be some performance implication as it would affect how the data and threads are mapped. I'm not sure what would be the most ideal way at this point, but would revisit if needed.